### PR TITLE
Add openfhe-julia

### PR DESCRIPTION
**Pull request by sloede**

### CODE REVIEW

**Comment by sloede** [comment]
@giordano Have you seen [this error from the Windows build](https://buildkite.com/julialang/yggdrasil/builds/7450#018cbe0c-5550-413f-98a8-c092b6a06de9/631-802) before:
```
[04:06:28]   The imported target "OPENFHEcore" references the file
[04:06:28] 
[04:06:28]      "/workspace/destdir/lib/libOPENFHEcore.dll"
[04:06:28] 
[04:06:28]   but this file does not exist.
```
When checking out the contents of autogenerated JLL tarballs for OpenFHE_jll (e.g., `OpenFHE.v1.1.2.x86_64-w64-mingw32-cxx11.tar.gz` from https://github.com/JuliaBinaryWrappers/OpenFHE_jll.jl/releases/tag/OpenFHE-v1.1.2%2B0), it seems like that indeed, all files in the `lib` directory have a `.dll.a` extension, while all the actual `.dll` files are in the `bin` directory.

Do you know what I can to to make the correct OpenFHE libraries available to downstream packages - do I need to modify the OpenFHE [`build_tarballs.jl`](https://github.com/JuliaPackaging/Yggdrasil/blob/master/O/OpenFHE/build_tarballs.jl) or is it something I can/need to fix here, e.g., by modifying my CMake setup?

**Comment by sloede** [comment]
Hm, looking at the file `OpenFHETargets-release.cmake` in the JLL product, I find the following lines:
```cmake
# Import target "OPENFHEcore" for configuration "Release"
set_property(TARGET OPENFHEcore APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(OPENFHEcore PROPERTIES
  IMPORTED_IMPLIB_RELEASE "${_IMPORT_PREFIX}/lib/libOPENFHEcore.dll.a"
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/libOPENFHEcore.dll"
  )

list(APPEND _IMPORT_CHECK_TARGETS OPENFHEcore )
list(APPEND _IMPORT_CHECK_FILES_FOR_OPENFHEcore "${_IMPORT_PREFIX}/lib/libOPENFHEcore.dll.a" "${_IMPORT_PREFIX}/lib/libOPENFHEcore.dll" )
```
Thus it seems clear to me that the `IMPORTED_LOCATION_RELEASE` path should refer to the **`bin`** directory and so should the path to the dll in `_IMPORT_CHECK_FILES_FOR_OPENFHEcore`.

To me, this strongly hints at a problem in the CMake files provided by OpenFHE_jll.jl. And indeed, fixing this manually in an interactive session of BB, the openfhe-julia build works. Do you, by any chance, know what could be wrong on the OpenFHE CMake setup side and what to do about it?

**Comment by sloede** [comment]
Ah, the BB auditor is at "fault" here:
```
<snip>
┌ Warning: lib/libOPENFHEbinfhe.dll should be in `bin`!
└ @ BinaryBuilder.Auditor ~/.julia/packages/BinaryBuilder/0CUml/src/Auditor.jl:229
┌ Warning: lib/libOPENFHEcore.dll should be in `bin`!
└ @ BinaryBuilder.Auditor ~/.julia/packages/BinaryBuilder/0CUml/src/Auditor.jl:229
┌ Warning: lib/libOPENFHEpke.dll should be in `bin`!
└ @ BinaryBuilder.Auditor ~/.julia/packages/BinaryBuilder/0CUml/src/Auditor.jl:229
┌ Warning: Simple buildsystem detected; Moving all `.dll` files to `bin`!
└ @ BinaryBuilder.Auditor ~/.julia/packages/BinaryBuilder/0CUml/src/Auditor.jl:240
```
is there a way I can prevent this? Or _should_ I even prevent this and instead manually modify the `OpenFHETargets-release.cmake` to reflect the changed paths?

**Comment by sloede** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/7880 attempts to hotfix the underlying reason for the error, let's hold off merging this PR until then.

**Comment by sloede** [comment]
@giordano ready to merge now 😊

**Comment by giordano** [comment]
I think you want to follow https://github.com/JuliaPackaging/Yggdrasil/blob/1210bc5fde582783d6dd87b5f36d4ad36fea42c8/L/libsingular_julia/build_tarballs.jl#L39-L44 https://github.com/JuliaPackaging/Yggdrasil/blob/1210bc5fde582783d6dd87b5f36d4ad36fea42c8/L/libsingular_julia/build_tarballs.jl#L53 so that you target all julia versions (libjulia breaks API/ABI across minor versions)

File: O/openfhe_julia/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+
```
Comment (giordano) [inline O/openfhe_julia/build_tarballs.jl:4]: https://github.com/JuliaPackaging/Yggdrasil/blob/1210bc5fde582783d6dd87b5f36d4ad36fea42c8/L/libsingular_julia/build_tarballs.jl#L6-L9
```suggestion

# See https://github.com/JuliaLang/Pkg.jl/issues/2942
# Once this Pkg issue is resolved, this must be removed
uuid = Base.UUID("a83860b7-747b-57cf-bf1f-3e79990d037f")
delete!(Pkg.Types.get_last_stdlibs(v"1.6.3"), uuid)

```

**Comment by sloede** [comment]
> I think you want to follow
> [...]
> so that you target all julia versions (libjulia breaks API/ABI across minor versions)

Thanks a lot for the hint (and the suggestion), that is incredibly helpful!

**Comment by sloede** [comment]
@giordano Do you have an idea why the binaries generated for, e.g., Julia v1.8 seem to work well ([GHA log](https://github.com/sloede/OpenFHE.jl/actions/runs/7380352504/job/20077626337)), while for v1.9 and v1.10 they fail with
```
ERROR: LoadError: InitError: could not load library "C:\Users\runneradmin\.julia\artifacts\dc58129f8d04ad71389779de34cc61fcb0618967\bin\libopenfhe_julia.dll"
%1 is not a valid Win32 application.
```
([GHA log](https://github.com/sloede/OpenFHE.jl/actions/runs/7380352504/job/20077626468#step:7:85))? I assume it has to do something with the libjulia_jll dependency, but I don't even know how to start debugging this 😢

**Pull request merged**